Source/Core/VideoBackends/Metal/MTLStateTracker.h

```diff
@@ -0,0 +1,275 @@
+
+  size_t Align(size_t amt, AlignMask align)
+  {
+    return (amt + static_cast<size_t>(align)) & ~static_cast<size_t>(align);
+  }
+  Map AllocateForTextureUpload(size_t amt);
+  Map Allocate(UploadBuffer buffer_idx, size_t amt, AlignMask align)
+  {
+    Preallocate(buffer_idx, amt);
+    return CommitPreallocation(buffer_idx, amt, align);
+  }
+  std::pair<void*, size_t> Preallocate(UploadBuffer buffer_idx, size_t amt);
+  /// Must follow a call to Preallocate where amt is >= to the one provided here
+  Map CommitPreallocation(UploadBuffer buffer_idx, size_t amt, AlignMask align)
+  {
+    assert((m_upload_buffers[static_cast<int>(buffer_idx)].usage.Pos() &
+            static_cast<size_t>(align)) == 0);
+    return CommitPreallocation(buffer_idx, Align(amt, align));
+  }
+  id<MTLBlitCommandEncoder> GetUploadEncoder();
+  id<MTLBlitCommandEncoder> GetTextureUploadEncoder();
+  id<MTLCommandBuffer> GetRenderCmdBuf();
+
+private:
+  class UsageTracker
+  {
+    struct UsageEntry
+    {
+      u64 drawno;
+      size_t pos;
+    };
+    std::vector<UsageEntry> m_usage;
+    size_t m_size = 0;
+    size_t m_pos = 0;
+
+  public:
+    size_t Size() { return m_size; }
+    size_t Pos() { return m_pos; }
+    bool PrepareForAllocation(u64 last_draw, size_t amt);
+    size_t Allocate(u64 current_draw, size_t amt);
+    void Reset(size_t new_size);
+  };
+
+  struct CPUBuffer
+  {
+    UsageTracker usage;
+    MRCOwned<id<MTLBuffer>> mtlbuffer;
+    void* buffer = nullptr;
+  };
+
+  struct BufferPair
+  {
+    UsageTracker usage;
+    MRCOwned<id<MTLBuffer>> cpubuffer;
+    MRCOwned<id<MTLBuffer>> gpubuffer;
+    void* buffer = nullptr;
+    size_t last_upload = 0;
+  };
+
+  struct Backref;
+  struct PerfQueryTracker;
+
+  std::shared_ptr<Backref> m_backref;
+  std::vector<std::shared_ptr<PerfQueryTracker>> m_perf_query_tracker_cache;
+  MRCOwned<id<MTLFence>> m_fence;
+  MRCOwned<id<MTLCommandBuffer>> m_upload_cmdbuf;
+  MRCOwned<id<MTLBlitCommandEncoder>> m_upload_encoder;
+  MRCOwned<id<MTLCommandBuffer>> m_texture_upload_cmdbuf;
+  MRCOwned<id<MTLBlitCommandEncoder>> m_texture_upload_encoder;
+  MRCOwned<id<MTLCommandBuffer>> m_current_render_cmdbuf;
+  MRCOwned<id<MTLCommandBuffer>> m_last_render_cmdbuf;
+  MRCOwned<id<MTLRenderCommandEncoder>> m_current_render_encoder;
+  MRCOwned<id<MTLComputeCommandEncoder>> m_current_compute_encoder;
+  MRCOwned<MTLRenderPassDescriptor*> m_render_pass_desc[3];
+  MRCOwned<MTLRenderPassDescriptor*> m_resolve_pass_desc;
+  Framebuffer* m_current_framebuffer;
+  CPUBuffer m_texture_upload_buffer;
+  BufferPair m_upload_buffers[static_cast<int>(UploadBuffer::Last) + 1];
+  u64 m_current_draw = 1;
+  std::atomic<u64> m_last_finished_draw{0};
+
+  MRCOwned<id<MTLTexture>> m_dummy_texture;
+
+  // MARK: State
+  u8 m_dirty_textures;
+  u8 m_dirty_samplers;
+  union Flags
+  {
+    struct
+    {
+      // clang-format off
+      bool has_gx_vs_uniform      : 1;
```

Comment:
`BitField` can be used here.

Reply:
I feel like the second best option here would be a flags enum like some of the other renderers use.  It definitely has its pluses, like being much easier to set/clear multiple values at once.

But I still went with this after seeing the messy, out of order state Vulkan's dirty flags are in, since no one wants to renumber all the flags just so they can add a new one in the middle.

It's definitely a choice I'm glad I went with during development, and while I don't expect to be adding new entries here nearly as often in the future, I'd still like to keep it easy.

Reply:
Yeah, the linter can be weird sometimes. My philosophy is to just write the code without thinking about the linter and just leave the formatting up to it when everything's done. You can leave it as is if you want.